src/DotNetty.Codecs/Base64/Base64.cs

```diff
@@ -0,0 +1,379 @@
+            }
+        }
+        private unsafe static int EncodeUsingGetSet(byte* alphabet, IByteBuffer src, IByteBuffer dest, int offset, int length, bool breakLines)
+        {
+            int srcReaderIndex = src.ReaderIndex;
+            int srcWriterIndex = src.WriterIndex;
+
+            int i = 0;
+            int j = 0;
+            int charCount = 0;
+            int calcLength = offset + length - length % 3;
+            for (i = offset; i < calcLength; i += 3)
+            {
+                if (breakLines)
+                {
+                    if (charCount == MAX_LINE_LENGTH)
+                    {
+                        dest.SetByte(j++, NEW_LINE);
+                        charCount = 0;
+                    }
+                    charCount += 4;
+                }
+
+                dest.SetByte(j + 0, alphabet[(src.GetByte(i) & 0xfc) >> 2]);
+                dest.SetByte(j + 1, alphabet[((src.GetByte(i) & 0x03) << 4) | ((src.GetByte(i + 1) & 0xf0) >> 4)]);
+                dest.SetByte(j + 2, alphabet[((src.GetByte(i + 1) & 0x0f) << 2) | ((src.GetByte(i + 2) & 0xc0) >> 6)]);
+                dest.SetByte(j + 3, alphabet[(src.GetByte(i + 2) & 0x3f)]);
+                j += 4;
+            }
+
+            i = calcLength;
+
+            if (breakLines && (length % 3 != 0) && (charCount == MAX_LINE_LENGTH))
+            {
+                dest.SetByte(j++, NEW_LINE);
+            }
+            switch (length % 3)
+            {
+                case 2:
+                    dest.SetByte(j + 0, alphabet[(src.GetByte(i) & 0xfc) >> 2]);
+                    dest.SetByte(j + 1, alphabet[((src.GetByte(i) & 0x03) << 4) | ((src.GetByte(i + 1) & 0xf0) >> 4)]);
+                    dest.SetByte(j + 2, alphabet[(src.GetByte(i + 1) & 0x0f) << 2]);
+                    dest.SetByte(j + 3, EQUALS_SIGN);
+                    j += 4;
+                    break;
+                case 1:
+                    dest.SetByte(j + 0, alphabet[(src.GetByte(i) & 0xfc) >> 2]);
+                    dest.SetByte(j + 1, alphabet[(src.GetByte(i) & 0x03) << 4]);
+                    dest.SetByte(j + 2, EQUALS_SIGN);
+                    dest.SetByte(j + 3, EQUALS_SIGN);
+                    j += 4;
+                    break;
+            }
+            //restore src's ReaderIndex and WriterIndex
+            src.SetIndex(srcReaderIndex, srcWriterIndex);
+            //remove last byte if it's NewLine
+            int destLength = dest.GetByte(j - 1) == NEW_LINE ? j - 1 : j;
+            return destLength;
+        }
+
+        public unsafe static IByteBuffer Encode(IByteBuffer src, int offset, int length, bool breakLines, Base64Dialect dialect, IByteBufferAllocator allocator)
+        {
+            if (src == null)
+            {
+                throw new ArgumentNullException(nameof(src));
+            }
+            if (dialect.alphabet == null)
+            {
+                throw new NullReferenceException(nameof(dialect.alphabet));
```

Comment:
>NullReferenceException [](start = 26, length = 22)

ArgumentNullException here and in other places where we check for "not null" and throw if it isn't so.